### DISCUSSION TITLE
🧩 d365fo.tools goes Power Platform

### DISCUSSION THREAD

**Discussion by FH-Inway** [d365collaborative/d365fo.tools#855] [Ideas]
Microsoft is currently transitioning hosting of D365FO environments from LCS to the Power Platform. This introduces a lot of changes that also impact d365fo.tools. If you are new to the topic or need a refresher, take a look at the following resources:
- [Unified admin experience](https://learn.microsoft.com/en-us/power-platform/admin/unified-experience/finance-operations-apps-overview)
- [Unified development experience](https://learn.microsoft.com/en-us/power-platform/developer/unified-experience/finance-operations-dev-overview)
- [One Dynamics One Platform TechTalk Series](https://community.dynamics.com/blogs/post/?postid=0bf71ad7-6f83-449e-8bb5-51672532f962)
- [Blog with roadmap](https://www.powerazure365.com/blog-1/dynamics-365-finops-unified-developer-experience#:~:text=AND%20FINALLY%20%2D%20REMEMBER%20IT%E2%80%99S%20JUST%20A%20MICROSOFT%20ROADMAP%20(SUBJECT%20TO%20CHANGE%E2%80%A6%20%3A)) by @AurelienClere-365

Inspired by #854 and the fact that I had opportunity to play with the unified experiences, I wanted to start this discussion on how we could get d365fo.tools ready for the Power Platform.

**Comment by FH-Inway** [comment]
At the moment, I just have a collection of random thoughts and ideas, in no particular order:
- (parts of) d365fo.tools can be used with unified environments. I was for example able to get [Get-D365Model](https://github.com/d365collaborative/d365fo.tools/blob/9a51364df3b061a6a07edd407de9264d6be04c0a/d365fo.tools/functions/get-d365model.ps1) and [Invoke-D365ModuleCompile](https://github.com/d365collaborative/d365fo.tools/blob/9a51364df3b061a6a07edd407de9264d6be04c0a/d365fo.tools/functions/invoke-d365modulecompile.ps1) to work on a Windows 11 vm with the development tools installed. See [d365fo.tools-unified.ps1](https://gist.github.com/FH-Inway/a01036a1939397e48cc2fe8e224a1074) for how this looks like.
- cmdlets similar to Get-D365Model (i.e. ones with `-BinDir` or other `Dir` parameters) can be used as is (though a bit painful)
  - there are however exceptions, e.g. [Invoke-D365BestPractice](https://github.com/d365collaborative/d365fo.tools/blob/9a51364df3b061a6a07edd407de9264d6be04c0a/d365fo.tools/functions/invoke-d365bestpractice.ps1) does not work, because there is no parameter to provide a directory for the `-packagesroot` parameter of `xppbp.exe`
- other cmdlets do not apply to the Power Platform, e.g. all the [LCS](https://github.com/search?q=repo%3Ad365collaborative%2Fd365fo.tools+lcs+path%3A%2F%5Ed365fo%5C.tools%5C%2Ffunctions%5C%2F%2F&type=code) cmdlets. 
- first step is probably to give d365fo.tools the capability to recognize if it is running on an environment with the unified development experience; the place to add this is probably in [variables.ps1](https://github.com/d365collaborative/d365fo.tools/blob/9a51364df3b061a6a07edd407de9264d6be04c0a/d365fo.tools/internal/scripts/variables.ps1#L64-L80); needs some research to determine a good criteria 
- next step would be some internal functions to determine the [current configuration](https://learn.microsoft.com/en-us/power-platform/developer/unified-experience/finance-operations-install-config-tools#:~:text=On%20the%20configuration%20form%2C%20select%20New%20and%20then%20create%20a%20new%20configuration.) and set some variables based on that
- open questions
  - in what order should cmdlets be made Power Platform ready? what are the most important ones? we currently have more than 200 cmdlets
  - how to handle cmdlets that cannot be made ready? 
  - how can the community help with this?

Looking forward to getting some feedback on this.

**Reply by AurelienClere-365** [reply]
Happy to help if needed 😉 - especially on the PAC CLI for F&O cmdlets

**Comment by FH-Inway** [comment]
On the topic of determining if d365fo.tools is being executed on a UDE: Seems that information of the active UDE configuration is stored in registry values at `Computer\HKEY_CURRENT_USER\Software\Microsoft\Dynamics\AX7\Development\Configurations`.

On local VHD and CHE environments, there is also a registry value `CompatibilityMode` set to `1` in `Computer\HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Dynamics\AX\Development`. According to @Pankaj-Thakur from Microsoft, this is a criteria for non-UDE environments (see https://www.yammer.com/dynamicsaxfeedbackprograms/threads/2817784834039808).

A related discussion is also happening here: https://github.com/ameyer505/D365FOAdminToolkit/issues/45#issuecomment-2408566210

## UDE registry
![image](https://github.com/user-attachments/assets/d5dbc12d-97c4-4319-9eee-808e0b751f09)

## CHE registry
![image](https://github.com/user-attachments/assets/66728fee-49fe-4b14-b2c5-069af31558d9)

**Comment by FH-Inway** [comment]
Based on the registry values, I started #868 to use those values for the d365fo.tools.

**Comment by FH-Inway** [comment]
I've been playing around with GitHub projects and the GitHub CLI / GraphQL API (see [GitHub-Project-Manager.ps1](https://gist.github.com/FH-Inway/9f1de243fbbdd612590935d0d46c8275)).

The idea is to have a project with one item per cmdlet of d365fo.tools. The project can then be used to organize and track the progress in making the module ready for a unified world. First draft of the project: https://github.com/users/FH-Inway/projects/2

A PoC script exists to populate the project with the cmdlet items. In the project draft, items 8 through 12 have been created with the script.

@Splaxi I made you a maintainer of the project draft, if you want to play around with it.

![image](https://github.com/user-attachments/assets/ef780ae1-dbb0-4d06-be3a-5edf65d4745b)
